sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/complexTypesSuite.scala

```diff
@@ -331,4 +330,31 @@ class ComplexTypesSuite extends PlanTest with ExpressionEvalHelper {
       .analyze
     comparePlans(Optimizer execute rel, expected)
   }
+
+  test("SPARK-23500: Simplify complex ops that aren't at the plan root") {
+    val structRel = relation
+      .select(GetStructField(CreateNamedStruct(Seq("att1", 'nullable_id)), 0, None) as "foo")
+      .groupBy($"foo")("1").analyze
+    val structExpected = relation
+      .select('nullable_id as "foo")
+      .groupBy($"foo")("1").analyze
+    comparePlans(Optimizer execute structRel, structExpected)
+
+    // If nullable attributes aren't used in the 'expected' plans, the array and map test
+    // cases fail because array and map indexing can return null so the output attribute

```

Comment:
This explains why the original plan(before optimize) marks its output as nullable, but I'm confused why the optimized plan still marks its output as nullable.

Reply:
It's a good question! I'm not too familiar with how nullability is marked and unmarked during planning. My understanding is roughly that the analyzer resolves all the plan's expressions and in doing so marks attributes as nullable or not. After that it's not clear that the optimizer revisits any of those nullability decisions. Is there an optimizer pass which should make nullability marking more precise?

Reply:
`nullable` is mostly calculated on demand, so we don't have rules to change the `nullable` property. For this case, the expression is `Alias(GetArrayItem(CreateArray(Attribute...)))`, which is nullable. After optimize, it becomes `Alias(Attribute...)` and is not nullable(if that attribute is not nullable). So the `nullable` is updated automatically.

I don't know why you hit this issue, please ping us if you can't figure it out, we can help to debug.

Reply:
Thanks, that's plenty of information to get started - I'll dig into it.

Reply:
@cloud-fan I looked again at this briefly this morning. The issue is that it's the `AttributeReference` in the top-level `Aggregate`'s `groupingExpressions` that has inconsistent nullability. 

The `AttributeReference` in the original plan was originally created with `nullable=true`, before optimization. So at that point it's kind of fixed unless the optimizer dereferences the attr reference and realises that the target is no longer nullable.

Reply:
Done, thanks. I filed SPARK-23634 to fix this. Out of interest, why does `AttributeReference` cache the nullability of its referent? Is it because comparison is too expensive to do if you have to follow a level of indirection to get to the original attribute?

Reply:
Because `AttributeReference` is not only used as a reference of an attribute from children, but also the new attributes produced by leaf nodes, which has to carry the nullable info. It's not ideal but it's too late to change now.